javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/extension/AgentListener.java

```diff
@@ -8,32 +8,51 @@
 import io.opentelemetry.instrumentation.api.cache.Cache;
 import io.opentelemetry.instrumentation.api.config.Config;
 import io.opentelemetry.instrumentation.api.field.VirtualField;
+import io.opentelemetry.sdk.autoconfigure.AutoConfiguredOpenTelemetrySdk;
 import java.lang.instrument.Instrumentation;
 import net.bytebuddy.agent.builder.AgentBuilder;
 
 /**
  * {@link AgentListener} can be used to execute code before/after Java agent installation, for
- * example to install any implementation providers that are used by instrumentations. For instance,
- * this project uses this SPI to install OpenTelemetry SDK.
+ * example to install any implementation providers that are used by instrumentations. Can also be
+ * used to obtain the {@link AutoConfiguredOpenTelemetrySdk}.
  *
  * <p>This is a service provider interface that requires implementations to be registered in a
  * provider-configuration file stored in the {@code META-INF/services} resource directory.
  */
 public interface AgentListener extends Ordered {
 
   /**
-   * Runs before the {@link AgentBuilder} construction, before any instrumentation is added.
+   * Runs before {@link AgentBuilder} construction, before any instrumentation is added.
    *
-   * <p>Execute only a minimal code because any classes loaded before the agent installation will
-   * have to be retransformed, which takes extra time, and more importantly means that fields can't
-   * be added to those classes - which causes {@link VirtualField} to fall back to the less
-   * performant {@link Cache} implementation for those classes.
+   * <p>Execute only minimal code because any classes loaded before the agent installation will have
+   * to be retransformed, which takes extra time, and more importantly means that fields can't be
+   * added to those classes - which causes {@link VirtualField} to fall back to the less performant
+   * {@link Cache} implementation for those classes.
    */
   default void beforeAgent(Config config) {}
```

Comment:
I considered deprecating this but `beforeAgent(AutoConfiguredOpenTelemetrySdk autoConfiguredOpenTelemetrySdk)` doesn't get called when there is a noop OpenTelemetry installed. If we think its ok to not have any agent listener hooks when noop is used, then I can go ahead and deprecate this because folks can access `AutoConfiguredOpenTelemetrySdk#getConfig()`.

Reply:
it would be really nice to deprecate this. not sure what to do about `noop`. @jkwatson @breedx-splk are you finding `noop` useful?

Reply:
I do think the noop is useful.

Over in [the original issue in the other repo](https://github.com/open-telemetry/opentelemetry-java/issues/3924) we chatted about the two arg `(config, sdk)` version being preferred. This two-version approach requires the SPI user to override both and hold state if they need both. It also looks weird (to me) at the call site below, because, depending on the implementation, `beforeAgent()` can be called twice.

My preference would be to just have the two-arg version and then later when the config problem (Config vs. ConfigProperties) is solved we can remove the config arg.

How's that sound?

Reply:
> My preference would be to just have the two-arg version and then later when the config problem (Config vs. ConfigProperties) is solved we can remove the config arg.

If we continue supporting `noop`, how would we remove the config arg once the config problem is solved? since in the `noop` case `AutoConfiguredOpenTelemetrySdk` is null, so it seems like we still need to pass around agent configuration somehow

Reply:
What about a noop implementation equivalent of the `AutoConfiguredOpenTelemetrySdk`?

Reply:
> What about a noop implementation equivalent of the `AutoConfiguredOpenTelemetrySdk`?

I like that idea 👍 
There are some instrumentations that are implemented as agent listener (e.g. metrics), or other features (like the profiler in splunk's case) that should be initialized even if the SDK is not initialized - so that we're able to measure the impact of these without the SDK. Not calling the listeners if noop OpenTelemetry is installed is kinda hacky and unreliable, and will produce wrong results in overhead tests.

How about implementing the listener this way:

```java
interface AgentListener {
  
  default void beforeAgent(OpenTelemetry openTelemetry, Resource resource, Config config) {
    beforeAgent(config);
  }

  @Deprecated
  default void beforeAgent(Config config) {}
  
  default void afterAgent(OpenTelemetry openTelemetry, Resource resource, Config config) {
    afterAgent(config);
  }

  @Deprecated
  default void afterAgent(Config config) {}
}
```

`AutoConfiguredOpenTelemetrySdk` contains `OpenTelemetrySdk` and we won't be able to fill that in with `OpenTelemetry.noop()`, so we can just pass its fields to the listener.

Reply:
> the config problem (Config vs. ConfigProperties)

About that: I think that currently I prefer using `Config` everywhere - it has all those get-or-default variants that will never throw and return the default value even when an error occurs during parsing. We don't want the agent to break the application because of some random property using wrong syntax after all.

Reply:
sound good 👍

Reply:
What if we keep `noop`, but only call `AgentListener` methods if the agent is enabled and the sdk is installed? Is there a use case for having the `AgentListener` when `noop` is installed? If we did that, we could define the `AgentListener` as:
```
interface AgentListener {
  
  default void beforeAgent(Config config, AutoConfiguredOpenTelemetrySdk autoConfiguredSdk) {
    beforeAgent(config);
  }

  @Deprecated
  default void beforeAgent(Config config) {}
  
  default void afterAgent(Config config, AutoConfiguredOpenTelemetrySdk autoConfiguredSdk) {
    afterAgent(config);
  }

  @Deprecated
  default void afterAgent(Config config) {}
}
```
I prefer having access to `AutoConfiguredOpenTelemetrySdk` versus `OpenTelemetry` and `Resource` because: 
- We have access to `OpenTelemetrySdk` instead of `OpenTelemetry`, which gives access to `SdkTracerProvider` and `SdkMeterProvider` instead of just `TracerProvider` and `MeterProvider`. This also means that we could give access to `SdkLogEmitterProvider` which will probably never be accessible via `OpenTelemetry`. 
- We give the user access to holder of all the autoconfigured things, which means that as the set of autoconfigured things changes, the user already access without further API changes.

Reply:
> Is there a use case for having the AgentListener when noop is installed?

this is a good question. I'm guessing ideally we would still install things like `RuntimeMetricsInstaller` in the `noop` case, but maybe we can live without that?

Reply:
> I prefer having access to `AutoConfiguredOpenTelemetrySdk` versus `OpenTelemetry` and `Resource` because:
> 
> * We have access to `OpenTelemetrySdk` instead of `OpenTelemetry`, which gives access to `SdkTracerProvider` and `SdkMeterProvider` instead of just `TracerProvider` and `MeterProvider`. This also means that we could give access to `SdkLogEmitterProvider` which will probably never be accessible via `OpenTelemetry`.
> * We give the user access to holder of all the autoconfigured things, which means that as the set of autoconfigured things changes, the user already access without further API changes.

Yeah, these are all perfectly valid reasons. Okay, consider me convinced -- I suppose we can live without running a few listeners in case of `noop`; all bytecode manipulations are still done, we're just not installing a couple of metrics.

Reply:
@trask if noop `OpenTelemetry` is installed then the runtime metrics won't have any affect, right?